src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncExceptionHandlerRewriter.cs

```diff
@@ -350,11 +363,32 @@ private BoundStatement UnpendBranches(
                     }
                 }
 
-                var caseStatement = _F.SwitchSection(i, unpendReturn);
-                cases.Add(caseStatement);
+                if (endOfBlockIsReachable)
+                {
+                    var caseStatement = _F.SwitchSection(i, unpendReturn);
+                    cases.Add(caseStatement);
+                }
+                else
+                {
+                    unconditionalReturn = unpendReturn;
+                }
             }
 
-            return _F.Switch(_F.Local(pendingBranchVar), cases.ToImmutableAndFree());
+            if (!endOfBlockIsReachable && unconditionalReturn is null && !_F.CurrentFunction.IsIterator)
+            {
+                // An example of this is a try block that unconditionally throws an exception.
+                // In this case, we just add `throw null;` to the end of the block to ensure that
+                // all possible cases are handled.
+                // We don't do this for iterators because their returns are handled within their own state
+                // machine, and adding this throw would simply cause an unconditional throw at the end of the finally,
+                // changing iterator behavior.
+                unconditionalReturn = _F.Throw(_F.Null(_F.SpecialType(SpecialType.System_Object)));
```

Comment:
From offline discussion, this was surprising. We could avoid this by turning the node from `UnpendException` into non-branching code (always rethrow)